pkg/ebpf/bytecode/runtime/printk_patcher.go

```diff
@@ -0,0 +1,218 @@
+// It's separated from PatchPrintkNewline so it can be tested independently, also so that we can
+// check how many patches are performed
+func patchPrintkInstructions(p *ebpf.ProgramSpec) (int, error) {
+	var errs []error // list of errors that happened while patching, if any
+	numPatches := 0  // number of patches performed
+
+	// Compute some opcodes we'll need
+	movImmOpCode := asm.Mov.Op(asm.ImmSource)
+	ldDWImmOpCode := asm.LoadImmOp(asm.DWord)
+	movRegOpCode := asm.Mov.Op(asm.RegSource)
+	addImmOpCode := asm.Add.Op(asm.ImmSource)
+
+	// In some cases the compiler might reuse the same instruction for multiple calls to bpf_trace_printk,
+	// keep track of that to avoid errors when we don't find a newline in an instruction we've already patched.
+	patchedInstructionIndexes := make(map[int]bool)
+
+	for idx, ins := range p.Instructions {
+		if !ins.IsBuiltinCall() || ins.Constant != int64(asm.FnTracePrintk) {
+			continue // Not a call to bpf_trace_printk, skip
+		}
+		maxLookback := max(0, idx-100) // For safety, don't look back more than 100 instructions
+
+		// We found the call to bpf_trace_printk, now we need to find
+		// the string on the stack and patch it.
+		// For that, find first the value of the second register, which is the second argument
+		// to the call, which is the length of the string.
+		// Example instruction: MovImm dst: r2 imm: 0x0000000000000009 which
+		// sets the length of the formatting to 9
+		var lengthLoadIns *asm.Instruction
+		for i := idx - 1; i >= maxLookback; i-- {
+			candidate := &p.Instructions[i]
+			if candidate.OpCode == movImmOpCode && candidate.Dst == asm.R2 {
+				lengthLoadIns = candidate
+				break
+			}
+		}
+		if lengthLoadIns == nil {
+			errs = append(errs, log.Warnf("Could not find length load instruction for bpf_trace_printk call %d in %s", idx, p.Name))
+			continue // Skip this call instruction
+		}
+
+		// Now we have to find in which part the stack is the string being stored
+		// For that we need to find the mov instruction that puts the stack pointer
+		// into r1 and then the add that modifies the stack offset
+		// We are looking for a sequence like this:
+		// MovReg dst: r1 src: rfp                 | Sets r1 to the stack pointer
+		// AddImm dst: r1 imm: 0x-000000000000048  | Adjusts the offset
+		var stackOffsetIns *asm.Instruction
+		for i := idx - 1; i >= maxLookback && stackOffsetIns == nil; i-- {
+			candidate := &p.Instructions[i]
+			if candidate.OpCode == movRegOpCode && candidate.Dst == asm.R1 && candidate.Src == asm.RFP {
+				// Ok, so we found the instruction that loads the stack pointer into r1
+				// From that, advance until we find the add instruction that modifies the stack offset
+				// (the AddImm instruction in the example above)
+				for j := i + 1; j < idx; j++ {
+					candidate = &p.Instructions[j]
+					if candidate.OpCode == addImmOpCode && candidate.Dst == asm.R1 {
+						stackOffsetIns = candidate
+						break
+					} else if candidate.Dst == asm.R1 {
```

Comment:
Conditional jumps can have one part of the condition in a register which is specified through the destination field of the instruction. We may not want to skip those, since they don't modify the register.
Example:
```
goto pc+8 if r1 == 9
```
See [this](https://github.com/cilium/ebpf/blob/main/asm/jump.go#L64) and [this](https://github.com/cilium/ebpf/blob/main/asm/dsl_test.go#L27).

Reply:
Thanks! I added a check to ensure class is ALU or load for this case, I think that should cover it right?

Reply:
Looks good!